networking_aci/plugins/ml2/drivers/mech_aci/allocations_manager.py

```diff
@@ -101,14 +108,15 @@ def _allocate_vlan_segment(self, network, host_id, level, host_config):
         return AllocationsModel(host=host_id, level=level, segment_type=segment_type, segmentation_id=segmentation_id,
                                 segment_id=segment.id, network_id=network_id)
 
+    @db_api.retry_db_errors
     def _allocate_vxlan_segment(self, network, host_id, level, host_config):
         LOG.info(_LI("Allocating segment for network type VXLAN"))
         segment_type = host_config.get('segment_type', 'vlan')
         segment_physnet = host_config.get('physical_network', None)
         network_id = network['id']
 
         session = db_api.get_writer_session()
-        with session.begin(subtransactions=True):
+        with db_api.exc_to_retry(db_exc.IntegrityError), session.begin(subtransactions=True):
```

Comment:
This block's function just converts any `IntegrityError` into a `RetryRequest` exception. The decorator of the function reacts to those `RetryRequest` exceptions and will retry the function.

Reply:
You're right. It's from `sqlalchemy`
```
neutron/db/db_base_plugin_v2.py:from sqlalchemy import exc as sql_exc
neutron/db/db_base_plugin_v2.py:        with db_api.exc_to_retry(sql_exc.IntegrityError),
```

Reply:
I don't get the second part. Do you mean, we should open the `with` later? That's not possible. It has to be around the `with session.begin()`, because that will automatically commit on exiting the block, which will then raise the `IntegrityError`.

Reply:
Makes sense, thanks.

Reply:
Good catch with the IntegrityError, fixed